toolchain/sem_ir/inst_fingerprinter.cpp

```diff
@@ -0,0 +1,338 @@
+    auto ir_inst = sem_ir->import_ir_insts().Get(ir_inst_id);
+    AddInFile(sem_ir->import_irs().Get(ir_inst.ir_id).sem_ir, ir_inst.inst_id);
+  }
+
+  template <typename T>
+    requires(std::same_as<T, BoolValue> ||
+             std::same_as<T, CompileTimeBindIndex> ||
+             std::same_as<T, ElementIndex> || std::same_as<T, FloatKind> ||
+             std::same_as<T, IntKind> || std::same_as<T, RuntimeParamIndex>)
+  auto Add(T arg) {
+    // Index-like ID: just include the value directly.
+    contents.push_back(arg.index);
+  }
+
+  template <typename T>
+    requires(std::same_as<T, AnyRawId> || std::same_as<T, ExprRegionId> ||
+             std::same_as<T, LocId> || std::same_as<T, RealId>)
+  auto Add(T /*arg*/) {
+    CARBON_FATAL("Unexpected instruction operand kind {0}", typeid(T).name());
+  }
+
+  // Add an instruction argument to the contents of the current instruction.
+  template <typename... Types>
+  auto AddWithKind(uint64_t arg, TypeEnum<Types...> kind) -> void {
+    using AddFunction = void (*)(Worklist& worklist, uint64_t arg);
+    using Kind = decltype(kind);
+
+    // Build a lookup table to add an argument of the given kind.
+    static constexpr std::array<AddFunction, Kind::NumTypes + 2> table = [] {
+      std::array<AddFunction, Kind::NumTypes + 2> table;
+      table[Kind::None.ToIndex()] = [](Worklist& /*worklist*/,
+                                       uint64_t /*arg*/) {};
+      table[Kind::Invalid.ToIndex()] = [](Worklist& /*worklist*/,
+                                          uint64_t /*arg*/) {
+        CARBON_FATAL("Unexpected invalid argument kind");
+      };
+      ((table[Kind::template For<Types>.ToIndex()] =
+            [](Worklist& worklist, uint64_t arg) {
+              return worklist.Add(Inst::FromRaw<Types>(arg));
+            }),
+       ...);
+      return table;
+    }();
+
+    table[kind.ToIndex()](*this, arg);
+  }
+
+  // Ensure all the instructions on the todo list have fingerprints.
+  auto Run() -> void {
+    while (!todo.empty()) {
+      auto [next_sem_ir, next_inst_id] = todo.back();
+      if (fingerprints->Contains(std::pair(next_sem_ir, next_inst_id))) {
+        todo.pop_back();
+        continue;
+      }
+
+      size_t init_size = todo.size();
+      auto inst = next_sem_ir->insts().Get(next_inst_id);
+      auto [arg0_kind, arg1_kind] = inst.ArgKinds();
+
+      // Prepare to fingerprint this instruction.
+      sem_ir = next_sem_ir;
+      contents.clear();
+
+      Add(inst.kind());
+
+      // Don't include the type if it's `type` or `<error>`, because those types
+      // are self-referential.
+      if (inst.type_id() != TypeType::SingletonTypeId &&
+          inst.type_id() != ErrorInst::SingletonTypeId) {
+        Add(inst.type_id());
+      }
+
+      for (auto [arg, kind] : {std::pair(inst.arg0(), arg0_kind),
+                               std::pair(inst.arg1(), arg1_kind)}) {
+        AddWithKind(arg, kind);
+      }
```

Comment:
This seems like two function calls would be a clearer way to write it, given it's just a list of two?
```suggestion
      AddWithKind(inst.arg0(), arg0_kind);
      AddWithKind(inst.arg1(), arg1_kind);
```

Reply:
Thanks! I used to have a big `switch` inside this loop but missed this cleanup when moving it out to a separate  function.